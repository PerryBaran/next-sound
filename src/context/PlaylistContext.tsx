'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
