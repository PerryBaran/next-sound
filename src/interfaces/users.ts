interface Songs {
  name: string
  url: string
  id: string
  createdAt: string
}

interface Albums {
  id: string
  name: string
  url: string
  UserId: string
  Songs: Songs[]
  createdAt: string
}

export default interface Users {
  id: string
  name: string
  email: string
  Albums: Albums[]
  createdAt: string
}
