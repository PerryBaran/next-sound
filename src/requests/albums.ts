import {
  postRequest,
  getRequest,
  patchRequest,
  deleteRequest,
  getByIdRequest
} from "./helpers"

export async function postAlbums(data: { name: string; image?: File }) {
  const response = await postRequest("albums", data)
  return response
}

export async function getAlbums(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("albums", query)
  return response
}

export async function getAlbumById(id: string) {
  const response = await getByIdRequest("albums", id)
  return response
}

export async function patchAlbum(
  id: string,
  data: { name?: string; image?: File }
) {
  const response = await patchRequest("albums", id, data)
  return response
}

export async function deleteAlbum(id: string) {
  const response = await deleteRequest("albums", id)
  return response
}
