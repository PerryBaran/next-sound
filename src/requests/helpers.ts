import Cookie from 'js-cookie'
import axios from 'axios'

const instance = axios.create({
  withCredentials: true,
  baseURL: 'https://soundcloud-clone-api.onrender.com'
})

function config() {
  return {
    headers: {
      userToken: Cookie.get('userToken')
    }
  }
}

type Model = 'users' | 'users/login' | 'users/signup' | 'albums' | 'string'

function createForm(model: string, data: any) {
  const formData = new FormData()
  switch (model) {
    case 'songs': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.audio) {
        formData.append('audio', data.audio)
      }
      if (data.position !== undefined) {
        formData.append('position', data.position)
      }
      if (data.AlbumId) {
        formData.append('AlbumId', data.AlbumId)
      }
      break
    }
    case 'albums': {
      if (data.name) {
        formData.append('name', data.name)
      }
      if (data.image) {
        formData.append('image', data.image)
      }
      break
    }
    default: {
      return data
    }
  }
  return formData
}

function error(err: any | { response: { data: { message: string } } }) {
  if (err.response?.data?.message) {
    throw new Error(err.response.data.message)
  }
  throw new Error('Unexpected Error')
}

export async function postRequest(model: Model, data: any) {
  const formData = createForm(model, data)

  try {
    const response = await instance.post(`/${model}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getRequest(
  model: Model,
  query:
    | {
        name?: string
        exact?: boolean
        limit?: number | string
      }
    | undefined
) {
  let endpoint = `/${model}`
  const queryArray = []
  if (query) {
    if (query.name) {
      queryArray.push(`name=${query.name}`)
      if (query.exact) {
        queryArray.push('exact=true')
      }
    }
    if (query.limit) {
      queryArray.push(`limit=${query.limit}`)
    }
    if (queryArray.length > 0) {
      const query = queryArray.join('&')
      endpoint = `${endpoint}?${query}`
    }
  }

  try {
    const response = await instance.get(endpoint, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function getByIdRequest(model: Model, id: string | number) {
  try {
    const response = await instance.get(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function patchRequest(
  model: Model,
  id: string | number,
  data: any
) {
  const formData = createForm(model, data)

  try {
    const response = await instance.patch(`/${model}/${id}`, formData, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}

export async function deleteRequest(model: Model, id: string | number) {
  try {
    const response = await instance.delete(`/${model}/${id}`, config())
    return response.data
  } catch (err: any | { response: { data: { message: string } } }) {
    error(err)
  }
}
