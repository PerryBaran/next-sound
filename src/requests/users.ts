import {
  postRequest,
  getRequest,
  getByIdRequest,
  patchRequest,
  deleteRequest
} from "./helpers"
import Cookies from "js-cookie"

export async function login(data: { email: string; password: string }) {
  const response = await postRequest("users/login", data)
  Cookies.set("userToken", response.userToken, { expires: 5 })
  return response
}

export async function signup(data: {
  name: string
  email: string
  password: string
}) {
  const response = await postRequest("users/signup", data)
  Cookies.set("userToken", response.userToken, { expires: 5 })
  return response
}

export async function getUsers(query?: {
  name?: string
  exact?: boolean
  limit?: number | string
}) {
  const response = await getRequest("users", query)
  return response
}

export async function getUserById(id: string) {
  const response = await getByIdRequest("users", id)
  return response
}

export async function patchUser(
  id: string,
  data: {
    name?: string
    email?: string
    password?: string
  }
) {
  const response = await patchRequest("users", id, data)
  return response
}

export async function deleteUser(id: string, password: string) {
  const response = await deleteRequest("users", `${id}/${password}`)
  return response
}
