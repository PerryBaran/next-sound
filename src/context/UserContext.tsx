'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
