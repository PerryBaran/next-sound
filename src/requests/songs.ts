import { postRequest, getRequest, patchRequest, deleteRequest } from "./helpers/requestHelpers"

export async function postSongs(data: {
  name: string
  audio: File
  position: number
  AlbumId: string
}) {
  const response = await postRequest("songs", data)
  return response
}

export async function getSongs(query?: {
  name?: string
  exact?: boolean
  limit?: number
}) {
  const response = await getRequest("songs", query)
  return response
}

export async function patchSong(
  id: string,
  data: { name?: string; audio?: File; position?: number }
) {
  const response = await patchRequest("songs", id, data)
  return response
}

export async function deleteSong(id: string) {
  const response = await deleteRequest("songs", id)
  return response
}
