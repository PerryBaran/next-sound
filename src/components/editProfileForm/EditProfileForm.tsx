"use client"

import { useUserContext } from "../../context/UserContext"
import { deleteUser, patchUser } from "../../requests/users"
import { useState } from "react"
import Alert from "../../components/alert/Alert"
import { useRouter } from "next/navigation"
import ConfirmPassword from "../../components/confirmPassword/ConfirmPassword"
import Confirm from "../../components/confirm/Confirm"
import css from "./editProfileForm.module.css"

export default function EditProfileForm({
  data,
  userId
}: {
  data: { name: string; email: string }
  userId: string
}) {
  const [userData, setUserData] = useState({
    name: data.name,
    email: data.email,
    password: "",
    confirmPassword: ""
  })
  const [alert, setAlert] = useState("")
  const router = useRouter()
  const { handleLogin } = useUserContext()
  const [confirm, setConfirm] = useState("")

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target
    setUserData((prev) => {
      return { ...prev, [name]: value }
    })
  }

  const handleSubmit = async () => {
    const EMAIL_REGEX = /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/

    if (
      userData.name === data.name &&
      userData.email === data.email &&
      !userData.password
    ) {
      setAlert("No changes requested")
    } else if (!userData.name) {
      setAlert("Name cannot be empty")
    } else if (
      userData.email !== data.email &&
      !userData.email.match(EMAIL_REGEX)
    ) {
      setAlert("Email must be valid")
    } else if (userData.password && userData.password.length < 8) {
      setAlert("Password must be atleast 8 characters long")
    } else if (
      userData.password &&
      userData.password !== userData.confirmPassword
    ) {
      setAlert("Passwords do not match")
    } else {
      const data = {
        name: userData.name || undefined,
        email: userData.email || undefined,
        password: userData.password || undefined
      }

      try {
        await patchUser(userId, data)

        if (userData.name) {
          handleLogin({ name: userData.name })
        }

        router.push(`/profile/${userData.name}`)
      } catch (err: any | Error) {
        setAlert(err?.message || "Unexpected Error")
      }
    }
  }

  const handleCancel = () => {
    router.back()
  }

  const handleDeleteAlbum = async (password: string) => {
    try {
      await deleteUser(userId, password)
      handleLogin()
      router.push("/")
    } catch (err: any | Error) {
      setAlert(err?.message || "Unexpected Error")
    }
  }

  const determineConfirm = (string: string) => {
    switch (string) {
      case "cancel":
        return handleCancel
      case "submit":
        return handleSubmit
      default:
        return () => {}
    }
  }

  return (
    <div className={css["container"]}>
      <Alert message={alert} />
      <form
        onSubmit={(e: React.FormEvent) => {
          e.preventDefault()
          setConfirm("submit")
        }}
        className={css["form"]}
      >
        <h2 className={css["heading"]}>Edit Profile</h2>
        <label htmlFor="name" className={css["field"]}>
          <span>Name</span>
          <input
            type="text"
            id="name"
            name="name"
            value={userData.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="email" className={css["field"]}>
          <span>Email</span>
          <input
            type="text"
            id="email"
            name="email"
            value={userData.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password" className={css["field"]}>
          <span>New Password</span>
          <input
            type="password"
            id="password"
            name="password"
            value={userData.password}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="confirm-password" className={css["field"]}>
          <span>Confirm New Password</span>
          <input
            type="password"
            id="confirm-password"
            name="confirmPassword"
            value={userData.confirmPassword}
            onChange={handleChange}
          />
        </label>
        <div className={css["buttons"]}>
          <button type="submit">Save Changes</button>
          <button type="button" onClick={() => setConfirm("delete")}>
            Delete Account
          </button>
          <button type="button" onClick={() => setConfirm("cancel")}>
            Cancel
          </button>
        </div>
      </form>
      {confirm === "delete" && (
        <ConfirmPassword callback={handleDeleteAlbum} setConfirm={setConfirm} />
      )}
      {confirm && confirm !== "delete" && (
        <Confirm setConfirm={setConfirm} callback={determineConfirm(confirm)} />
      )}
    </div>
  )
}
