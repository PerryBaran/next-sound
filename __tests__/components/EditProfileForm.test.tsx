import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { createContext, useContext } from "react"
import EditProfileForm from "../../src/components/editProfileForm/EditProfileForm"
import * as userRequests from "../../src/requests/users"
import * as navigation from "next/navigation"
import * as UserContext from "../../src/context/UserContext"

describe("EditProfile", () => {
  let deleteUserSpy: jest.SpyInstance
  let patchUserSpy: jest.SpyInstance
  let mockPush = jest.fn()
  let mockBack = jest.fn()
  const mockData = {
    name: "user-name",
    email: "email@email.com"
  }
  let mockHandleLogin = jest.fn()
  const mockUser = {
    name: mockData.name,
    id: "user-id"
  }

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(() => {
    mockPush = jest.fn()
    mockBack = jest.fn()
    jest.spyOn(navigation, "useRouter").mockReturnValue({
      back: mockBack,
      forward: jest.fn(),
      push: mockPush,
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn()
    })

    deleteUserSpy = jest.spyOn(userRequests, "deleteUser").mockResolvedValue("")
    patchUserSpy = jest.spyOn(userRequests, "patchUser")

    mockHandleLogin = jest.fn()
    const mockContext = createContext({
      user: mockUser,
      handleLogin: mockHandleLogin
    })
    jest
      .spyOn(UserContext, "useUserContext")
      .mockImplementation(() => useContext(mockContext))
  })

  describe("tests", () => {
    beforeEach(() => {
      render(<EditProfileForm data={mockData} userId={mockUser.id} />)
    })

    test("renders correctly", () => {
      expect(screen.getByText(/edit profile/i)).toBeInstanceOf(
        HTMLHeadingElement
      )
      expect(screen.getByLabelText(/name/i)).toHaveAttribute("type", "text")
      expect(screen.getByLabelText(/email/i)).toHaveAttribute("type", "text")
      expect(screen.getByLabelText(/^new password$/i)).toHaveAttribute(
        "type",
        "password"
      )
      expect(screen.getByLabelText(/confirm new password/i)).toHaveAttribute(
        "type",
        "password"
      )
      expect(screen.getByText(/save changes/i)).toHaveAttribute(
        "type",
        "submit"
      )
      expect(screen.getByText(/delete account/i)).toHaveAttribute(
        "type",
        "button"
      )
      expect(screen.getByText(/cancel/i)).toHaveAttribute("type", "button")
    })

    test("cancel button", () => {
      expect(mockBack).toBeCalledTimes(0)

      fireEvent.click(screen.getByText(/cancel/i))
      fireEvent.click(screen.getByText(/yes/i))

      expect(mockBack).toBeCalledTimes(1)
    })

    test("delete button", async () => {
      expect(deleteUserSpy).toBeCalledTimes(0)
      expect(mockHandleLogin).toBeCalledTimes(0)
      expect(mockPush).toBeCalledTimes(0)

      const password = "password"

      fireEvent.click(screen.getByText(/delete account/i))
      fireEvent.change(screen.getByLabelText(/confirm password/i), {
        target: { value: password }
      })
      fireEvent.click(screen.getByText(/^confirm$/i))

      await waitFor(() => {
        expect(deleteUserSpy).toBeCalledWith(mockUser.id, password)
        expect(mockHandleLogin).toBeCalledWith()
        expect(mockPush).toBeCalledWith("/")
      })
    })

    describe("submit button", () => {
      test("no changes", () => {
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        expect(screen.getByText(/no changes requested/i)).toBeInTheDocument()
      })

      test("empty name", () => {
        fireEvent.change(screen.getByLabelText(/name/i), {
          target: { value: "" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        expect(screen.getByText(/name cannot be empty/i)).toBeInTheDocument()
      })

      test("invalid email", () => {
        fireEvent.change(screen.getByLabelText(/email/i), {
          target: { value: "1234" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        expect(screen.getByText(/email must be valid/i)).toBeInTheDocument()
      })

      test("password too short", () => {
        fireEvent.change(screen.getByLabelText(/^new password$/i), {
          target: { value: "1234" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        expect(
          screen.getByText(/Password must be atleast 8 characters long/i)
        ).toBeInTheDocument()
      })

      test("new passwords don't match", () => {
        fireEvent.change(screen.getByLabelText(/^new password$/i), {
          target: { value: "12345678" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        expect(screen.getByText(/passwords do not match/i)).toBeInTheDocument()
      })

      test("succesfully updates", async () => {
        const data = {
          name: "new-name",
          email: "new-email@email.com",
          password: "new-password"
        }
        patchUserSpy.mockResolvedValue(data)

        expect(patchUserSpy).toBeCalledTimes(0)
        expect(mockHandleLogin).toBeCalledTimes(0)
        expect(mockPush).toBeCalledTimes(0)

        fireEvent.change(screen.getByLabelText(/name/i), {
          target: { value: data.name }
        })
        fireEvent.change(screen.getByLabelText(/email/i), {
          target: { value: data.email }
        })
        fireEvent.change(screen.getByLabelText(/^new password$/i), {
          target: { value: data.password }
        })
        fireEvent.change(screen.getByLabelText(/confirm new password/i), {
          target: { value: data.password }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        await waitFor(() => {
          expect(patchUserSpy).toBeCalledWith(mockUser.id, data)
          expect(mockHandleLogin).toBeCalledWith({ name: data.name })
          expect(mockPush).toBeCalledWith(`/profile/${data.name}`)
        })
      })

      test("throws expected error", async () => {
        const error = "error"
        patchUserSpy.mockRejectedValue({ message: error })

        fireEvent.change(screen.getByLabelText(/name/i), {
          target: { value: "new-name" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        await waitFor(() => {
          expect(screen.getByText(error)).toBeInTheDocument()
        })
      })

      test("throws unexpected error", async () => {
        patchUserSpy.mockRejectedValue("")

        fireEvent.change(screen.getByLabelText(/name/i), {
          target: { value: "new-name" }
        })
        fireEvent.click(screen.getByText(/save changes/i))
        fireEvent.click(screen.getByText(/yes/i))

        await waitFor(() => {
          expect(screen.getByText(/unexpected error/i)).toBeInTheDocument()
        })
      })
    })
  })

  test("snapshot", () => {
    const { asFragment } = render(
      <EditProfileForm data={mockData} userId={mockUser.id} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
