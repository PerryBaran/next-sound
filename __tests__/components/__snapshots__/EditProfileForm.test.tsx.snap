// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfile snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class="form"
    >
      <h2
        class="heading"
      >
        Edit Profile
      </h2>
      <label
        class="field"
        for="name"
      >
        <span>
          Name
        </span>
        <input
          id="name"
          name="name"
          type="text"
          value="user-name"
        />
      </label>
      <label
        class="field"
        for="email"
      >
        <span>
          Email
        </span>
        <input
          id="email"
          name="email"
          type="text"
          value="email@email.com"
        />
      </label>
      <label
        class="field"
        for="password"
      >
        <span>
          New Password
        </span>
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
      </label>
      <label
        class="field"
        for="confirm-password"
      >
        <span>
          Confirm New Password
        </span>
        <input
          id="confirm-password"
          name="confirmPassword"
          type="password"
          value=""
        />
      </label>
      <div
        class="buttons"
      >
        <button
          type="submit"
        >
          Save Changes
        </button>
        <button
          type="button"
        >
          Delete Account
        </button>
        <button
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
